src/main/java/com/laytonsmith/core/functions/CompositeFunction.java

```diff
@@ -54,6 +61,7 @@ public final Mixed exec(Target t, Environment environment, Mixed... args) throws
 		} else {
 			tree = CACHED_SCRIPTS.get(this.getClass());
 		}
+		setTarget(tree, t);
```

Comment:
Hmm, this will cause issues if there are multithreaded uses of the function, so I think this is not a good idea to do this. Additionally, the setTarget function is going to be quite inefficient, because every call requires O(n) processing of the whole tree. I'd rather just catch an exception that the eval throws, and then set the target accordingly. Then we are limited to changing things that only affect this thread, and we avoid having to loop through all the nodes in the tree.

Reply:
Nice idea

Reply:
Actually, this is more complicated, because the exception stack trace needs to be modified, at the correct spot, because lower stack entries may already have a target that's correct, so it's not a matter of just setting the correct target on the exception itself, it's about setting the target of the stack trace element. I would need to dig further into this to figure out exactly how to do that, but if you want to do it, you'll need to examine the exception's stack trace (which is actually stored in the environment, not the exception), and make the appropriate change there.